docs/development/pigs/pig-009.rst

```diff
@@ -0,0 +1,298 @@
+
+InverseCDFSampler
+-----------------
+
+The first building block of the event sampler is a class that samples from a given 
+probability mass function (PMF). The proposed implementation follows closely what
+is described here_:
+
+
+The basic design of the class is as following:
+
+.. code::
+
+    from gammapy.utils.distributions import InverseCDFSampler
+
+    pdf = np.array([])
+    cdf_sampler = InverseCDFSampler(pdf=pdf, random_state=random_state)
+
+Where the `gammapy.utils.random.get_random_state` method is used. The class 
+implements mainly the `.sample()` method to draw samples from the given pdf:
+
+.. code::
+
+    cdf_sampler = InverseCDFSampler()
+    samples = cdf_sampler.sample(size=100)
+
+The returned object is a `np.ndarray` containing the sampled indices.
+
+In addition the `InverseCDFSampler` should have the possibility to sample only 
+along a given axis of the PDF. This features is required to handle sampling of 
+multiple PDFs at the same time to avoid a Python loop over the PDFs. This will 
+be supported by specifying the `axis` argument:
+
+.. code::
+
+    cdf_sampler = InverseCDFSampler(pdf=pdfs, axis=0)
+    samples = cdf_sampler.sample(size=1)
+
+
+The `InverseCDFSampler` could addtionally implement a few helper functions to plot
+the CDF and PDF. The `InverseCDFSampler` class will replace the current
+`GeneralRandom` and `GeneralRandomArray` classes in `gammapy.utils.distributions`. 
+
+
+
+MapEventSampler
+---------------
+
+To support sampling from the predicted number of counts maps for source as well 
+as background maps, we propose to introduce a `MapEventSampler` class. The 
+responsibility of this class is to handle the physical coordinate transformations, 
+time dependence of the model, output data structures and sampling of the total 
+number of observed events.
+Starting from a predicted number of counts map (e.g. computed by the `MapEvaluator`):
+
+.. code::
+
+    src = SkyModel()
+    evaluator = MapEvaluator(src, exposure, psf=None, edisp=None)
+    npred_map = evaluator.compute_npred()
+
+The `MapEventSampler` takes this map and a lightcurve model as an input and 
+allows to sample from it with the `.sample()` method:
+
+.. code::
+
+    lightcurve = LightCurveTableModel() # or PhaseCurveTableModel()
+    src_sampler = MapEventSampler(npred_map, lightcurve, t_min=, t_max=)
+
+    # draw total number of events
+    n_events = src_sampler.npred_total()
+
+    # draw events as an astropy table
+    events_src = map_sampler.sample()
+
+If no lightcurve is provided a constant rate can be assumed e.g. for 
+background models:
+
+.. code::
+
+    bkg_map = BackgroundModel().map
+    bkg_sampler = MapEventSampler(bkg_map, t_min=, t_max=)
+    events_bkg = bkg_sample.sample()
+
+
+The `MapEventSampler` returns an `astropy.table.Table` object, with the following
+columns: `TIME`, `RA_TRUE`, `DEC_TRUE`, `ENERGY_TRUE`. Later additional info such
+as `DETX` and `DETY` could be added.
+
+
+IRFEventDistributor
```

Comment:
Now I'm confused.

The MapEventSampler works on predicted counts maps, no?
In Gammapy we compute those with PSF and EDISP already applied.
What you get is RECO coordinates and energy directly by sampling that.

Only now when reading your proposal it dawns on me that actually you propose that the `MapEventSampler` is based on `npred = exposure * flux` maps without PSF and EDISP applied. Is that right?

If that is the proposal I would suggest to change the organisation of the proposal part: first explain the big picture, then drill down and explain the pieces. The big picture is that you first compute `npred = exposure * flux` and sample that, then adjust RA / DEC / ENERGY according to the PSF and EDISP, and then compute TIME from the 1D time dependence?

Some questions that should become clear in the overview:
- does the order of PSF and EDISP matter? Are those separate steps or coupled?
- TIME is completely independent, could also be computed first? The reason to do `npred` map first would be to have the total number of events known? Is that correct, or does TIME sampling depend on PSF / EDISP in any way?